ykcompile/src/lib.rs

```diff
@@ -850,73 +871,179 @@ impl<TT> TraceCompiler<TT> {
                     0 => (), // ZST.
                     1 => {
                         dynasm!(self.asm
-                            ; mov BYTE [Rq(ro.reg) + ro.offs], Rb(reg)
+                            ; mov BYTE [Rq(dest_ro.reg) + dest_ro.offs], Rb(src_reg)
                         );
                     }
                     2 => {
                         dynasm!(self.asm
-                            ; mov WORD [Rq(ro.reg) + ro.offs], Rw(reg)
+                            ; mov WORD [Rq(dest_ro.reg) + dest_ro.offs], Rw(src_reg)
                         );
                     }
                     4 => {
                         dynasm!(self.asm
-                            ; mov DWORD [Rq(ro.reg) + ro.offs], Rd(reg)
+                            ; mov DWORD [Rq(dest_ro.reg) + dest_ro.offs], Rd(src_reg)
                         );
                     }
                     8 => {
                         dynasm!(self.asm
-                            ; mov QWORD [Rq(ro.reg) + ro.offs], Rq(reg)
+                            ; mov QWORD [Rq(dest_ro.reg) + dest_ro.offs], Rq(src_reg)
                         );
                     }
-                    _ => todo!(),
+                    _ => unreachable!(),
+                }
+            }
+            (Location::Mem(dest_ro), Location::Mem(src_ro)) => {
+                if ty.size() <= 8 {
+                    let temp = self.create_temporary();
+                    match ty.size() {
+                        0 => (), // ZST.
+                        1 => {
+                            dynasm!(self.asm
+                                ; mov Rb(temp), BYTE [Rq(src_ro.reg) + src_ro.offs]
+                                ; mov BYTE [Rq(dest_ro.reg) + dest_ro.offs], Rb(temp)
+                            );
+                        }
+                        2 => {
+                            dynasm!(self.asm
+                                ; mov Rw(temp), WORD [Rq(src_ro.reg) + src_ro.offs]
+                                ; mov WORD [Rq(dest_ro.reg) + dest_ro.offs], Rw(temp)
+                            );
+                        }
+                        4 => {
+                            dynasm!(self.asm
+                                ; mov Rd(temp), DWORD [Rq(src_ro.reg) + src_ro.offs]
+                                ; mov DWORD [Rq(dest_ro.reg) + dest_ro.offs], Rd(temp)
+                            );
+                        }
+                        8 => {
+                            dynasm!(self.asm
+                                ; mov Rq(temp), QWORD [Rq(src_ro.reg) + src_ro.offs]
+                                ; mov QWORD [Rq(dest_ro.reg) + dest_ro.offs], Rq(temp)
+                            );
+                        }
+                        _ => unreachable!(),
+                    }
+                    self.free_if_temp(Location::Register(temp, false));
+                } else {
+                    self.c_memcpy(dest_ro, src_ro, ty.size());
```

Comment:
Can these locations overlap?

Reply:
And thus should we use `memmove()`. Good question.

I don't think they can overlap. @ptersilie?

Reply:
Can you guys think of a rust assignment that would copy into memory which overlaps with the source operand? Does the borrow checker disallow that perhaps.

Reply:
The MIR semantics regarding aliasing are not clear right now.

Reply:
I honestly don't know. But gut feeling says no, but who knows?

Reply:
It sounds like if the buffers overlap, we invoke undefined behaviour. So to be completely safe, I suppose we should use `memmove` and take the performance penalty.

Reply:
Yeah I agree that's the sensible approach.